### PR TITLE
[POU 49] - Crediting filter and date selector for Search Bar

### PR DESCRIPTION
**Pull request by MingLiang-PALOIT**
JIRA Ticket: POU-49

Description (Optional):
The date selector has been built with Dayzed components and rendered in Chakra for better alignment with design

### CODE REVIEW

File: src/@types/Calendar.tsx [hunk 1]
```diff
@@ -0,0 +1,2 @@
+export const MONTH_NAMES_SHORT = [`Jan`, `Feb`, `Mar`, `Apr`, `May`, `Jun`, `Jul`, `Aug`, `Sep`, `Oct`, `Nov`, `Dec`]
+export const WEEKDAY_NAMES_SHORT = [`Sun`, `Mon`, `Tue`, `Wed`, `Thu`, `Fri`, `Sat`]
```
Comment (morphet81) [inline src/@types/Calendar.tsx:2]: Need to get it automatically for current locale, else need to localize it

File: src/@types/SearchFilterValues.ts [hunk 2]
```diff
@@ -4,4 +4,10 @@ export interface SearchFilterValues {
   methodologies: string[]
   sectors: string[]
   countries: string[]
+  filterDates?: FilterDates
+}
+
+export interface FilterDates {
```
Comment (morphet81) [inline src/@types/SearchFilterValues.ts:10]: `DateFilter` seems better, don't you think?

File: src/components/molecules/CalendarBody/CalendarBody.tsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+interface CalendarBodyProps {
+  renderProps: RenderProps
+  nextMonth: () => void
+  previousMonth: () => void
+  nextYear: () => void
+  previousYear: () => void
+}
+
+const CalenderBody: FC<CalendarBodyProps> = (props) => {
```
Comment (morphet81) [inline src/components/molecules/CalendarBody/CalendarBody.tsx:16]: Typooooo

File: src/components/molecules/CalendarBodyDates/CalendarBodyDates.tsx [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            background = selected ? `lightGray.800` : background
+            background = !selectable ? `transparent` : background
+            return (
+              <WrapItem key={key} boxSize={`48px`} justifyContent={`center`}>
+                {selectable ? (
+                  <Button variant={`calendarDate`} color={selected ? `white` : `black`} backgroundColor={background} {...getDateProps({ dateObj })}>
+                    {date.getDate()}
+                  </Button>
+                ) : null}
+              </WrapItem>
```
Comment (morphet81) [inline src/components/molecules/CalendarBodyDates/CalendarBodyDates.tsx:43]: You can map over a list and genearte a list of react components outside of the return. Would be cleaner. Have some examples online.

File: src/components/molecules/CalendarBodyDates/CalendarBodyDates.tsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  return (
+    <Box>
+      <Wrap spacing={0}>
+        {WEEKDAY_NAMES_SHORT.map((weekday) => (
+          <WrapItem key={`${calendar.month}${calendar.year}${weekday}`} boxSize={`48px`} justifyContent={`center`}>
+            <Text textAlign={`center`} fontFamily={`body`} fontWeight={`medium`} margin={`auto`}>
+              {weekday[0]}
+            </Text>
+          </WrapItem>
+        ))}
```
Comment (morphet81) [inline src/components/molecules/CalendarBodyDates/CalendarBodyDates.tsx:23]: You can map over a list and genearte a list of react components outside of the return. Would be cleaner. Have some examples online.

File: src/components/molecules/CalendarWrapper/CalendarWrapper.tsx [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  }
+
+  const handleOnOKClick = () => {
+    applySelectedDate(selectedDate)
+  }
+
+  return (
+    <VStack>
+      <Dayzed
+        onDateSelected={(selectedDate, event) => setSelectedDate(selectedDate.date)}
```
Comment (morphet81) [inline src/components/molecules/CalendarWrapper/CalendarWrapper.tsx:55]: Naaah!!! Already told you numerous times to have function for handling events. And name it `handleSomething`

File: src/components/molecules/CalendarWrapper/CalendarWrapper.tsx [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        showOutsideDays
+        offset={offset}
+        maxDate={maxDate}
+        minDate={minDate}
+        selected={selectedDate}
+        render={(renderProps) => (
+          <CalenderBody renderProps={renderProps} nextMonth={handleNextMonth} previousMonth={handlePreviousMonth} nextYear={handleNextYear} previousYear={handlePreviousYear} />
+        )}
```
Comment (morphet81) [inline src/components/molecules/CalendarWrapper/CalendarWrapper.tsx:63]: Same here, not an event, but still a function.

File: src/components/molecules/CreditingPeriodFilter/CreditingPeriodFilter.tsx [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+  const renderLabel = () => {
+    if (minimumDate && maximumDate) {
+      return `${format(minimumDate, `dd/MM/yyyy`)} - ${format(maximumDate, `dd/MM/yyyy`)}`
+    }
+    if (minimumDate) {
+      return `${format(minimumDate, `dd/MM/yyyy`)} and later`
+    }
+    if (maximumDate) {
+      return t(`upTo`, { date: format(maximumDate, `dd/MM/yyyy`) })
+      // return `Up to ${format(maximumDate, `dd/MM/yyyy`)}`
```
Comment (morphet81) [inline src/components/molecules/CreditingPeriodFilter/CreditingPeriodFilter.tsx:103]: Forgot this one

File: src/components/molecules/CreditingPeriodFilter/CreditingPeriodFilter.tsx [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+            {t(`apply`)}
+          </Button>
+        </HStack>
+      </VStack>
+    )
+  }
+
+  const renderLabel = () => {
+    if (minimumDate && maximumDate) {
+      return `${format(minimumDate, `dd/MM/yyyy`)} - ${format(maximumDate, `dd/MM/yyyy`)}`
```
Comment (morphet81) [inline src/components/molecules/CreditingPeriodFilter/CreditingPeriodFilter.tsx:96]: Date patterns shouldn't be hard-corded. Please use the date patterns enum Mederic should have created.

Comment (morphet81) [reply src/components/molecules/CreditingPeriodFilter/CreditingPeriodFilter.tsx:96]: And please change that for all hard-coded date pattern in your files.

File: src/components/molecules/DateInput/DateInput.spec.tsx [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    const tOnChange = jest.fn()
+    const toOenDatePicker = jest.fn()
+
+    const { container } = render(<DateInput label={tLabel} value={undefined} onChange={tOnChange} openDatePicker={toOenDatePicker} maxDate={tDate} />)
+
+    await user.type(screen.getByRole(`textbox`), `02/09/2023`)
+    await user.tab()
+    expect(container).toMatchSnapshot()
+  })
+  test(`renders a DateInput and click on the Calendar icon`, async () => {
```
Comment (morphet81) [inline src/components/molecules/DateInput/DateInput.spec.tsx:100]: Break line. Break break break. Break. Always break.

File: src/components/molecules/DateInput/DateInput.tsx [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+      setTextInputValue(value.toLocaleDateString(`en-sg`))
+    } else {
+      setTextInputValue(``)
+    }
+  }, [value])
+
+  const inputConfig = {
+    height: `28px`,
+    padding: 0,
+    _active: { boxShadow: `none` },
```
Comment (morphet81) [inline src/components/molecules/DateInput/DateInput.tsx:32]: Why do you use underscores in names?

File: src/components/molecules/DateInput/DateInput.tsx [hunk 12]
```diff
@@ -0,0 +1,87 @@
+import { ChangeEvent, FC, useEffect, useState } from 'react'
+import { isValid, parse } from 'date-fns'
+import { Input, InputGroup, InputRightElement, Text, VStack } from '@chakra-ui/react'
+import { CalendarIcon } from '@/components/atoms/CalendarIcon/CalendarIcon'
+
+interface DateInputProp {
+  label: string
+  openDatePicker: () => void
```
Comment (morphet81) [inline src/components/molecules/DateInput/DateInput.tsx:8]: `onOpenDatePicker`. Generally, all event handlers passed as props should be named starting with `on`

File: src/components/organisms/SearchBar/SearchBar.tsx [hunk 13]
```diff
@@ -76,6 +77,7 @@ const SearchBar: FC = () => {
             noOfSelectedFilters={selectedSearchFilters.countries.length}
             applyFilters={setCountriesFilter}
           />
+          <CreditingPeriodFilter label={`Crediting Period`} />
```
Comment (morphet81) [inline src/components/organisms/SearchBar/SearchBar.tsx:80]: i18n

**Review by morphet81 — CHANGES_REQUESTED** [review]

File: src/@types/Calendar.tsx (hunk 1 above)
Comment (MingLiang-PALOIT) [reply src/@types/Calendar.tsx]: got it, found a way to generate them using date-fns

File: src/components/molecules/DateInput/DateInput.tsx (hunk 11 above)
Comment (MingLiang-PALOIT) [reply src/components/molecules/DateInput/DateInput.tsx:32]: ah, these are how Chakra deals with pseudo classes

File: src/components/molecules/CalendarBodyDates/CalendarBodyDates.tsx (hunk 4 above)
Comment (MingLiang-PALOIT) [reply src/components/molecules/CalendarBodyDates/CalendarBodyDates.tsx]: have extracted the generation of the React components out

File: src/components/molecules/CalendarWrapper/CalendarWrapper.tsx (hunk 6 above)
Comment (MingLiang-PALOIT) [reply src/components/molecules/CalendarWrapper/CalendarWrapper.tsx]: have extracted the generation of the React components out

File: src/components/molecules/CalendarBodyDates/CalendarBodyDates.tsx (hunk 5 above)
Comment (MingLiang-PALOIT) [reply src/components/molecules/CalendarBodyDates/CalendarBodyDates.tsx]: have extracted the generation of the React components out

File: src/components/molecules/CalendarBody/CalendarBody.tsx [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+import CalendarNavigation from '@/components/molecules/CalendarNavigation/CalendarNavigation'
+import { getMonthNames } from '@/@types/Calendar'
+import CalendarBodyDates from '@/components/molecules/CalendarBodyDates/CalendarBodyDates'
+
+interface CalendarBodyProps {
+  renderProps: RenderProps
+  nextMonth: () => void
+  previousMonth: () => void
+  nextYear: () => void
+  previousYear: () => void
```
Comment (morphet81) [inline src/components/molecules/CalendarBody/CalendarBody.tsx:13]: Should start with `on`

File: src/components/molecules/CalendarBodyDates/CalendarBodyDates.tsx [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+            const key = `${calendar.month}${calendar.year}${weekIndex}${index}`
+            if (!dateObj) {
+              return <WrapItem key={key} width={`calc(100% / 7)`}></WrapItem>
+            }
+            return renderCalendarDates(dateObj, key)
+          }),
+        )}
+      </>
+    )
+  }
```
Comment (morphet81) [inline src/components/molecules/CalendarBodyDates/CalendarBodyDates.tsx:30]: That's not what I meant. You should be able to create a new array from a `array.map` that returns react component and just use this array in rendering.

File: src/components/molecules/CalendarWrapper/CalendarWrapper.tsx [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+import { useTranslation } from 'react-i18next'
+import CalendarBody from '@/components/molecules/CalendarBody/CalendarBody'
+import { differenceInCalendarMonths } from 'date-fns'
+import { MONTHS_IN_YEAR } from '@/@types/Calendar'
+
+interface CalendarWrapperProps {
+  maxDate?: Date
+  minDate?: Date
+  preSelectedDate?: Date
+  applySelectedDate: (date: Date | undefined) => void
```
Comment (morphet81) [inline src/components/molecules/CalendarWrapper/CalendarWrapper.tsx:14]: Should start with `on`

File: src/components/molecules/CreditingPeriodFilter/CreditingPeriodFilter.tsx [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+    setMaximumDate(selectedSearchFilterValues.searchFilterValues.creditingPeriod?.maxDate)
+  }, [selectedSearchFilterValues.searchFilterValues.creditingPeriod])
+
+  const renderDateSelector = (stateValue: Date | undefined, action: Dispatch<SetStateAction<Date | undefined>>) => {
+    return (
+      <CalendarWrapper
+        preSelectedDate={stateValue}
+        maxDate={latestDate}
+        minDate={earliestDate}
+        applySelectedDate={(selectedDate) => {
```
Comment (morphet81) [inline src/components/molecules/CreditingPeriodFilter/CreditingPeriodFilter.tsx:41]: No inline function

**Review by morphet81 — CHANGES_REQUESTED** [review]

File: src/components/molecules/CalendarBodyDates/CalendarBodyDates.tsx [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+      <Wrap spacing={0}>
+        {weekDayNames.map((weekday) => (
+          <WrapItem key={`${calendar.month}${calendar.year}${weekday}`} boxSize={`48px`} justifyContent={`center`}>
+            <Text textAlign={`center`} fontFamily={`body`} fontWeight={`medium`} margin={`auto`}>
+              {weekday[0]}
+            </Text>
+          </WrapItem>
+        ))}
+      </Wrap>
+      <Wrap spacing={0}>{dateElements.map((element) => element)}</Wrap>
```
Comment (morphet81) [inline src/components/molecules/CalendarBodyDates/CalendarBodyDates.tsx:61]: Hum... nope. You are basically recreating `dateElements` from `dateElements`. So what's the best optimizedway?

File: src/components/molecules/CalendarWrapper/CalendarWrapper.tsx [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+        render={(renderProps) => (
+          <CalendarBody
+            renderProps={renderProps}
+            onNextMonth={handleOnNextMonth}
+            onPreviousMonth={handleOnPreviousMonth}
+            onNextYear={handleOnNextYear}
+            onPreviousYear={handleOnPreviousYear}
+          />
+        )}
+      />
```
Comment (morphet81) [inline src/components/molecules/CalendarWrapper/CalendarWrapper.tsx:73]: Raaaah!!!!

File: src/components/molecules/CreditingPeriodFilter/CreditingPeriodFilter.tsx [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+          />
+        </HStack>
+        <HStack justify={`space-between`} width={`100%`} margin={`8px`}>
+          <Button
+            variant={`textLink`}
+            onClick={() => {
+              setMinimumDate(undefined)
+              setMaximumDate(undefined)
+              setCreditingPeriodFilter({})
+            }}
```
Comment (morphet81) [inline src/components/molecules/CreditingPeriodFilter/CreditingPeriodFilter.tsx:84]: Rahhhh!

File: src/components/molecules/CreditingPeriodFilter/CreditingPeriodFilter.tsx [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+            onOpenDatePicker={() => setShowMinimumDateSelector(true)}
+            maxDate={latestDate}
+            minDate={earliestDate}
+            onChange={setMinimumDate}
+          />
+          <MinusIcon />
+          <DateInput
+            label={`Maximum Date`}
+            value={maximumDate}
+            onOpenDatePicker={() => setShowMaximumDateSelector(true)}
```
Comment (morphet81) [inline src/components/molecules/CreditingPeriodFilter/CreditingPeriodFilter.tsx:71]: No inline

File: src/components/molecules/CreditingPeriodFilter/CreditingPeriodFilter.tsx [hunk 22]
```diff
@@ -0,0 +1,137 @@
...
+    if (showMaximumDateSelector || showMinimumDateSelector) {
+      return null
+    }
+    return (
+      <VStack>
+        <HStack width={`100%`} justify={`center`} alignItems={`end`}>
+          <DateInput
+            label={`Minimum Date`}
+            value={minimumDate}
+            onOpenDatePicker={() => setShowMinimumDateSelector(true)}
```
Comment (morphet81) [inline src/components/molecules/CreditingPeriodFilter/CreditingPeriodFilter.tsx:62]: No inline

File: src/components/molecules/CreditingPeriodFilter/CreditingPeriodFilter.tsx [hunk 23]
```diff
@@ -0,0 +1,137 @@
...
+            onClick={() => {
+              setMinimumDate(undefined)
+              setMaximumDate(undefined)
+              setCreditingPeriodFilter({})
+            }}
+            data-testid={`crediting-period-filter-clear`}
+          >
+            {t(`clear`)}
+          </Button>
+          <Button variant={`textLink`} onClick={() => setCreditingPeriodFilter({ maxDate: maximumDate, minDate: minimumDate })} data-testid={`crediting-period-filter-apply`}>
```
Comment (morphet81) [inline src/components/molecules/CreditingPeriodFilter/CreditingPeriodFilter.tsx:89]: No inline

**Review by morphet81 — CHANGES_REQUESTED** [review]

**Review by morphet81 — APPROVED** [review]

**Pull request merged**